### PR TITLE
Default application_name for bgworker

### PR DESCRIPTION
**Pull request by weinan003**
At present, gpstop check pg_stat_activity table before stop cluster in
smart model. However, for those background workers which invoked BackgroundWorkerInitializeConnection(),
It maintains a session to cluster until the process stop. Hence, gpstop
does not work in smart model if these background workers active.

Give a default application_name for bgworker as a hint for gpstop. If
bgworker does not change the guc, gpstop will ignore these processes in
session check step.

Further context :
https://groups.google.com/a/greenplum.org/forum/#!topic/gpdb-dev/XS3K9nwDQFA

### CODE REVIEW

**Comment by weinan003** [comment]
> How does this work in upstream PostgreSQL? Is `pg_ctl -m smart` affected by background workers that are started similarly?

I glanced at pg11.

At present, Greenplum smart mode behavior is different from PostgreSQL pg_ctl.
pg_ctl -m smart mode sends "SIGTERM" signal directly and waits for all background workers quit.

While for gpstop, it just cancels the stop job if "pg_stat_active" table has any tuple fetched.

**Comment by Eulerizeit** [comment]
@weinan003  to me this seems like an issue with the way we do gpstop.  I'm not sure how easy it would be to address the issue in the current gpstart/gpstop code but from my perspective I'd like gpstop to behave as closely to pg_ctl as we can. 

@bharath1028 @jmcatamney @jchampio could one of you take a look and offer an opinion?

**Comment by jchampio** [comment]
My $0.02: this seems like the wrong way to go, but the right way to go also seems like it might be a lot of work for 6. We should be deferring these sorts of checks to the server itself (since the server understands what is and is not a user connection) but if we don't coordinate that check between all segments at the same time, it's possible that we'll end up with a half-stopped cluster.

If we *do* go this route, we at least need to use an `application_name` that's ugly enough that an end user wouldn't choose it accidentally. For example, the one of the hacks in place for gpstate uses `__gprecoverseg_pg_rewind__` as its `application_name`.

**Comment by weinan003** [comment]
> If we _do_ go this route, we at least need to use an `application_name` that's ugly enough that an end user wouldn't choose it accidentally. For example, the one of the hacks in place for gpstate uses `__gprecoverseg_pg_rewind__` as its `application_name`.

@jchampio  The `application_name` is just set by background worker not by user manually. This is to keep Greenplum and Postgres have the same stop behavior, i.e the smart stop should not block by any background worker processes. Greenplum has some auxiliary processes like `global deadlock detector`, they also have connections to a database. But they have not appeared in pg_stat_activities. We could also avoid background workers to appear in pg_stat_activities. But this is not consistent with upstream. Since in Postgres, background workers also appear in pg_stat_activities.

So we add a `bgworker` flag for background workers in pg_stat_activities to help gpstop to know they are not user connections and smart stop should not be blocked.
Is that make sense?

**Comment by asimrp** [comment]
What if we skip checking `pg_stat_activity` altogether from within `gpstop`?  That is, to make `gpstop` simpler and rely on postmaster.c's method to wait for backends depending on shutdown mode.  All the intelligence is already present in server, according to `pmdie()` where different shutdown modes are handled.

**Comment by jchampio** [comment]
> What if we skip checking pg_stat_activity altogether from within gpstop?

@asimrp I think that's what @Eulerizeit and I are hoping for as an eventual solution. I'm just worried that removing the (admittedly poor) up-front check might make some common use cases much worse until we replace that logic with something else.

> The application_name is just set by background worker not by user manually.

@weinan003 `application_name` is an end-user feature; there's nothing stopping someone from running
```
$ psql "port=15432 dbname=postgres application_name=bgworker"
psql (9.4.20, server 8.3.23)
Type "help" for help.

postgres=# select application_name from pg_stat_activity ;
 application_name
------------------
 bgworker
(1 row)
```

**Comment by weinan003** [comment]
> > What if we skip checking pg_stat_activity altogether from within gpstop?
> 
> @asimrp I think that's what @Eulerizeit and I are hoping for as an eventual solution. 
How about that I remove the check first so background worker will not impact `gpstop` works properly
@jchampio @asimrp

**Comment by HaozhouWang** [comment]
My 2 cents is that can we use application_name as a flag to notify the `gpstop` whether this connection could be canceled or not. Therefore, we can set the bg_worker's application_name like `%can_be_canceled%`. Then `gpstop` will skip the connections with such flags. Also, users can use this flag to indicate this connection which can be canceled and no need to wait in cluster stopping process

**Comment by HaozhouWang** [comment]
@jchampio @jmcatamney @dkrieger-pvtl, please review
Thank you very much!

File: gpMgmt/bin/gppylib/db/catalog.py [hunk 1]
```diff
@@ -36,9 +36,13 @@ def getDatabaseList(conn):
     sql = "SELECT datname FROM pg_catalog.pg_database"
     return basicSQLExec(conn,sql)
 
-def getUserPIDs(conn):
+def getUserPIDs(conn, ignoreLst):
     """dont count ourselves"""
     sql = """SELECT pid FROM pg_stat_activity WHERE pid != pg_backend_pid()"""
+    if ignoreLst != []:
+        ignoreStr = ', '.join("'"+ i +"'" for i in ignoreLst)
```
Comment (jchampio) [inline gpMgmt/bin/gppylib/db/catalog.py]: If `ignoreLst` can contain arbitrary strings, `i` needs to be run through an escaping layer to avoid accidental SQL injection. `pygresql.pg` provides one:
```python
from pygresql import pg
...
    ignoreStr = ', '.join("'%s'" % pg.escape_string(i) for i in ignoreLst)
```

File: gpMgmt/bin/gpstop [hunk 2]
```diff
@@ -37,6 +37,13 @@ try:
 except ImportError, e:
     sys.exit('ERROR: Cannot import modules.  Please check that you have sourced greenplum_path.sh.  Detail: ' + str(e))
 
+#GpstopCheckIgnoreList is a write list for pg_stat_active table check
```
Comment (jchampio) [inline gpMgmt/bin/gpstop]: Did you mean "[whitelist](https://en.wikipedia.org/wiki/Whitelisting)"?

File: gpMgmt/bin/gpstop [hunk 3]
```diff
@@ -37,6 +37,13 @@ try:
...
+#
+#Before stop cluster, gpstop smart model do sanity check depends on pg_stat_active table
+#to promise no session alive. However, some background workers may want to be ignored.
+#For this propose, we maintain a write list to indicate those `application_name` can be skipped checking.
```
Comment (dkrieger-pvtl) [inline gpMgmt/bin/gpstop]: This comment needs some grammatical touchup:
```
# Before stopping the cluster, gpstop smart mode does sanity checking based on the pg_stat_activity table
# to ensure that no session is alive. However, this check should ignore some background workers.
# We maintain a whitelist of strings, which will be matched against pg_stat_activity.application_name
```

File: gpMgmt/bin/gppylib/db/catalog.py [hunk 4]
```diff
@@ -36,9 +36,13 @@ def getDatabaseList(conn):
     sql = "SELECT datname FROM pg_catalog.pg_database"
     return basicSQLExec(conn,sql)
 
-def getUserPIDs(conn):
+def getUserPIDs(conn, ignoreLst):
```
Comment (dkrieger-pvtl) [inline gpMgmt/bin/gppylib/db/catalog.py]: nit: I prefer `ignoreList` for this variable name(here  as `ignoreLst` makes me wonder what `Lst` means).

File: gpMgmt/bin/gpstop [hunk 5]
```diff
@@ -104,6 +111,7 @@ class GpStop:
         self.timeout = timeout
         self.logfileDirectory = logfileDirectory
         self.onlyThisHost = onlyThisHost
+        self.checkIgnoreLst = checkIgnoreLst
```
Comment (dkrieger-pvtl) [inline gpMgmt/bin/gpstop]: nit: rename `checkIgnoreLst` to `checkIgnoreList`.

File: gpMgmt/bin/gpstop [hunk 6]
```diff
@@ -37,6 +37,13 @@ try:
 except ImportError, e:
     sys.exit('ERROR: Cannot import modules.  Please check that you have sourced greenplum_path.sh.  Detail: ' + str(e))
 
+#GpstopCheckIgnoreList is a write list for pg_stat_active table check
+#
+#Before stop cluster, gpstop smart model do sanity check depends on pg_stat_active table
+#to promise no session alive. However, some background workers may want to be ignored.
+#For this propose, we maintain a write list to indicate those `application_name` can be skipped checking.
+GpstopCheckIgnoreList = ['gpdiskquota']
```
Comment (dkrieger-pvtl) [inline gpMgmt/bin/gpstop]: Do we prefer a more obviously reserved name other than `gpdiskquota`?  Maybe `gp_internal_use_gpdiskquota`?

File: gpMgmt/bin/gpstop [hunk 7]
```diff
@@ -37,6 +37,15 @@ try:
 except ImportError, e:
     sys.exit('ERROR: Cannot import modules.  Please check that you have sourced greenplum_path.sh.  Detail: ' + str(e))
 
+"""
+GpstopCheckIgnoreList is a whitelist for pg_stat_active table check
```
Comment (dkrieger-pvtl) [inline gpMgmt/bin/gpstop]: I'd rephrase this to specify the intent `GpstopCheckIgnoreList is a whitelist to ignore when checking active database connections `

File: gpMgmt/bin/gpstop [hunk 8]
```diff
@@ -37,6 +37,15 @@ try:
...
+
+In smart model, we check `pg_stat_activity` table to promise no session alive before cluster stop routine.
+However, some background workers' life cycle are same as the cluster, so they want to be ignored in this step.
+For this propose, we maintain a whitelist to indicate those `application_name` can be skipped checking.
+"""
```
Comment (dkrieger-pvtl) [inline gpMgmt/bin/gpstop]: I'd rephrase the grammar here:
```
In smart mode, we check the `pg_stat_activity` table for  active sessions before calling the cluster stop routine.
However, we need to ignore those that are part of the database implementation itself.
For this propose, we maintain a whitelist using the `application_name` field.
```

File: gpMgmt/bin/gpstop [hunk 9]
```diff
@@ -37,6 +37,15 @@ try:
...
+GpstopCheckIgnoreList = ['gp_reserved_gpdiskquota']
```
Comment (ashwinstar) [inline gpMgmt/bin/gpstop]: It seems extremely strange for logic to depend on something,  `gp_reserved_gpdiskquota` in this case, which appears no where in that repository.

Comment (dkrieger-pvtl) [reply gpMgmt/bin/gpstop]: I presumed that this will part of a separate commit at a later time, but that they wanted to separate these changes from these.

Comment (weinan003) [reply gpMgmt/bin/gpstop]: Yes, `gp_reserved_gpdiskquota` will set in `gpdiskquota` (https://github.com/greenplum-db/diskquota) after this pr merged.

File: gpMgmt/bin/gpstop [hunk 10]
```diff
@@ -37,6 +37,16 @@ try:
 except ImportError, e:
     sys.exit('ERROR: Cannot import modules.  Please check that you have sourced greenplum_path.sh.  Detail: ' + str(e))
 
+"""
+GpstopCheckIgnoreList is a whitelist to ignore when checking active database connections
+
+In smart mode, we check the `pg_stat_activity` table for active sessions before calling
+the cluster stop routine. However, we need to ignore those that are part of the database
+implementation itself. For this propose, we maintain a whitelist using its `application_name`
```
Comment (dkrieger-pvtl) [inline gpMgmt/bin/gpstop]: Sorry...there was a typo in my comment below...'propose' should be 'purpose'.
